Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,141 @@
+import enum
+
+from types import TracebackType
+from typing import final, Any, Dict, Optional, Type
+
+from . import events
+from . import tasks
+
+
+__all__ = (
+    "Timeout",
+    "timeout",
+    "timeout_at",
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    EXPIRING = "expiring"
+    EXPIRED = "expired"
+    EXITED = "finished"
+
+
+@final
+class Timeout:
+
+    def __init__(self, deadline: Optional[float]) -> None:
```

Comment:
Is there any use case for deadline = None?

Reply:
deadline is disabled

Reply:
I suggest to remove None as a valid deadline as that is not useful. One this which would be valuable is to reject or cancel the current timeout.

Reply:
> I suggest to remove None as a valid deadline as that is not useful. 

I disagree. `aiohttp` has a good story for using `None` for disabling (skipping) timeout context managers.

> One this which would be valuable is to reject or cancel the current timeout.

Sorry, I don't follow. Please elaborate on what do you mean by *reject or cancel the current timeout*.

Reply:
> Sorry, I don't follow. Please elaborate on what do you mean by reject or cancel the current timeout.

There could be "cancel" method to cancel the current timeout like this:

```py
async with asyncio.timeout(1) as t:
    result = await <expr>
    if result:
        t.cancel() # Now timeout is disabled
    await <more expr>
```

But since there is already a reschedule method which can disable the timeout, this might be unnecessary.

Reply:
Oh, I think I see the use case that Andrew is talking about. Say there's a function that gets passed some work and a deadline. It does
```py
def foo(work, deadline):
    async with timeout(deadline):
        work()  # In reality this would have more code around it
```
If the caller doesn't want a deadline they can pass None as the deadline, and the timeout handler will just disable itself. Any other solution would complicate the code in foo() to special-case the "no deadline please" case.

Reply:
@gvanrossum thanks, it is precisely my use case.